File: tests/units/PSWSMan.Authentication.Tests/Acceptor.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;

namespace PSWSMan.Authentication.Tests;

/// <summary>The result of an acceptor step.</summary>
/// <param name="Token">The token to feed back to the client, null when the acceptor has nothing to send.</param>
/// <param name="Complete">Whether the acceptor considers the context established.</param>
internal sealed record AcceptorStepResult(byte[]? Token, bool Complete);

/// <summary>The result of the acceptor wrapping data for WinRM.</summary>
internal sealed record AcceptorWinRMWrapResult(byte[] Header, byte[] Data, int PaddingLength);

/// <summary>The password credentials a CredSSP acceptor received from the client.</summary>
internal sealed record AcceptorDelegatedCredentials(string Domain, string Username, string Password);

/// <summary>What the acceptor reports about its context.</summary>
internal sealed record AcceptorContextInfo(bool Complete, string? NegotiatedProtocol, int ContextAttr,
    string? ClientPrincipal, string? TlsProtocol, string? TlsCipher,
    AcceptorDelegatedCredentials? DelegatedCredentials);

/// <summary>An NTLM user the acceptor will accept.</summary>
internal sealed record AcceptorUser(string Domain, string Username, string Password);

/// <summary>Raised when the acceptor reports a failure for an operation.</summary>
internal sealed class AcceptorException(string type, string message) : Exception($"{type}: {message}")
{
    /// <summary>The Python exception type name, e.g. BadBindingsError.</summary>
    public string Type { get; } = type;
}

/// <summary>
/// A pyspnego acceptor running in a child Python process, driven over stdin/stdout with one JSON object per line. See
/// acceptor.py for the protocol.
/// </summary>
internal sealed class Acceptor : IDisposable
{
    private static readonly JsonSerializerOptions s_jsonOptions = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    private readonly Process _process;
    private readonly StreamWriter _stdin;
    private readonly StreamReader _stdout;
    private readonly StringBuilder _stderr = new();
    private readonly string? _userFile;
    private bool _disposed;

    private Acceptor(Process process, string? userFile)
    {
        _process = process;
        _stdin = process.StandardInput;
        _stdout = process.StandardOutput;
        _userFile = userFile;

        process.ErrorDataReceived += (_, e) =>
        {
            if (e.Data is not null)
            {
                lock (_stderr)
                {
                    _stderr.AppendLine(e.Data);
                }
            }
        };
        process.BeginErrorReadLine();
    }

    /// <summary>Starts the acceptor process, skipping the current test when Python with pyspnego is unavailable.</summary>
    /// <param name="users">The NTLM users to accept, written to the NTLM_USER_FILE the pure Python acceptor reads.</param>
    public static Acceptor Start(params AcceptorUser[] users)
    {
        string python = PythonHost.Require();

        string? userFile = null;
        if (users.Length > 0)
        {
            userFile = Path.Combine(Path.GetTempPath(), $"pswsman-ntlm-{Guid.NewGuid():N}.txt");
            StringBuilder content = new();
            foreach (AcceptorUser user in users)
            {
                content.Append(user.Domain).Append(':').Append(user.Username).Append(':').Append(user.Password)
                    .Append('\n');
            }
            File.WriteAllText(userFile, content.ToString());
        }

        ProcessStartInfo psi = new(python)
        {
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            WorkingDirectory = AppContext.BaseDirectory,
        };
        psi.ArgumentList.Add(Path.Combine(AppContext.BaseDirectory, "acceptor.py"));
        psi.Environment["PYTHONIOENCODING"] = "utf-8";
        psi.Environment["PYTHONUNBUFFERED"] = "1";
        if (userFile is not null)
        {
            psi.Environment["NTLM_USER_FILE"] = userFile;
        }

        Process process = Process.Start(psi) ?? throw new InvalidOperationException("Failed to start acceptor");
        return new Acceptor(process, userFile);
    }

    /// <summary>Creates the acceptor context.</summary>
    /// <param name="protocol">The pyspnego protocol, ntlm, kerberos, negotiate or credssp.</param>
    /// <param name="options">pyspnego NegotiateOptions names controlling the implementation used.</param>
    /// <param name="hostname">The acceptor host name used to build its SPN.</param>
    /// <param name="service">The acceptor service used to build its SPN.</param>
    /// <param name="channelBindings">The application data of the channel bindings the acceptor expects.</param>
    /// <param name="tls">CredSSP only, pins the acceptor's TLS version and cipher family, see acceptor.py.</param>
    public void Create(string protocol, string[]? options = null, string hostname = "unspecified",
        string service = "host", byte[]? channelBindings = null, string? tls = null)
    {
        Send(new
        {
            op = "create",
            protocol,
            options,
            hostname,
            service,
            channel_bindings = channelBindings,
            tls,
        });
    }

    public AcceptorStepResult Step(byte[]? token)
    {
        JsonElement response = Send(new { op = "step", token });
        return new AcceptorStepResult(Bytes(response, "token"), response.GetProperty("complete").GetBoolean());
    }

    public byte[] Wrap(byte[] data)
        => Bytes(Send(new { op = "wrap", data }), "data")!;

    public byte[] Unwrap(byte[] data)
        => Bytes(Send(new { op = "unwrap", data }), "data")!;

    public AcceptorWinRMWrapResult WrapWinRM(byte[] data)
    {
        JsonElement response = Send(new { op = "wrap_winrm", data });
        return new AcceptorWinRMWrapResult(
            Bytes(response, "header")!,
            Bytes(response, "data")!,
            response.GetProperty("padding_length").GetInt32());
    }

    public byte[] UnwrapWinRM(byte[] header, byte[] data)
        => Bytes(Send(new { op = "unwrap_winrm", header, data }), "data")!;

    public AcceptorContextInfo Query()
    {
        JsonElement response = Send(new { op = "query" });

        AcceptorDelegatedCredentials? delegated = null;
        if (response.TryGetProperty("delegated_credentials", out JsonElement creds))
        {
            delegated = new AcceptorDelegatedCredentials(
                creds.GetProperty("domain").GetString() ?? "",
                creds.GetProperty("username").GetString() ?? "",
                creds.GetProperty("password").GetString() ?? "");
        }

        return new AcceptorContextInfo(
            response.GetProperty("complete").GetBoolean(),
            response.GetProperty("negotiated_protocol").GetString(),
            response.GetProperty("context_attr").GetInt32(),
            response.GetProperty("client_principal").GetString(),
            response.TryGetProperty("tls_protocol", out JsonElement tlsProtocol) ? tlsProtocol.GetString() : null,
            response.TryGetProperty("tls_cipher", out JsonElement tlsCipher) ? tlsCipher.GetString() : null,
            delegated);
    }

    private JsonElement Send(object request)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        _stdin.Write(JsonSerializer.Serialize(request, s_jsonOptions));
        _stdin.Write('\n');
        _stdin.Flush();

        string? line = _stdout.ReadLine();
        if (line is null)
        {
            _process.WaitForExit(5000);
            throw new InvalidOperationException(
                $"Acceptor exited with {(_process.HasExited ? _process.ExitCode : "no exit code")} while waiting for a response.\n{Stderr}");
        }

        JsonElement response = JsonDocument.Parse(line).RootElement;
        if (!response.GetProperty("ok").GetBoolean())
        {
            throw new AcceptorException(
                response.GetProperty("type").GetString() ?? "Exception",
                response.GetProperty("error").GetString() ?? "");
        }

        return response;
    }

    private static byte[]? Bytes(JsonElement element, string property)
    {
        JsonElement value = element.GetProperty(property);
        return value.ValueKind == JsonValueKind.Null ? null : value.GetBytesFromBase64();
    }

    private string Stderr
    {
        get
        {
            lock (_stderr)
            {
                return _stderr.ToString();
            }
        }
    }

    public void Dispose()
    {
        if (_disposed)
        {
            return;
        }
        _disposed = true;

        try
        {
            _stdin.Write("{\"op\":\"exit\"}\n");
            _stdin.Flush();
            _stdin.Close();
            if (!_process.WaitForExit(5000))
            {
                _process.Kill();
            }
        }
        catch (IOException)
        {
            // The process already went away.
        }
        finally
        {
            _process.Dispose();
            if (_userFile is not null)
            {
                File.Delete(_userFile);
            }
        }
    }
}

/// <summary>Locates the Python interpreter that has pyspnego installed.</summary>
internal static class PythonHost
{
    /// <summary>
    /// The environment variable the build script sets to the interpreter of the virtual environment it provisions.
    /// Without it the interpreters on PATH are probed.
    /// </summary>
    public const string EnvironmentVariable = "PSWSMAN_TEST_PYTHON";

    private static readonly Lazy<string?> s_python = new(Locate, LazyThreadSafetyMode.ExecutionAndPublication);

    /// <summary>The interpreter path, or null when none with pyspnego was found.</summary>
    public static string? Path => s_python.Value;

    /// <summary>Returns the interpreter path or skips the current test when there is none.</summary>
    public static string Require()
    {
        string? python = Path;
        if (python is null)
        {
            Skip.Test($"Python with pyspnego is not available, set {EnvironmentVariable} or run build.ps1 -Task Test");
        }

        return python!;
    }

    private static string? Locate()
    {
        List<string> candidates = [];
        string? configured = Environment.GetEnvironmentVariable(EnvironmentVariable);
        if (!string.IsNullOrWhiteSpace(configured))
        {
            candidates.Add(configured);
        }
        candidates.Add("python3");
        candidates.Add("python");

        foreach (string candidate in candidates)
        {
            if (HasSpnego(candidate))
            {
                return candidate;
            }
        }

        return null;
    }

    private static bool HasSpnego(string python)
    {
        try
        {
            ProcessStartInfo psi = new(python)
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
            };
            psi.ArgumentList.Add("-c");
            psi.ArgumentList.Add("import spnego");

            using Process? process = Process.Start(psi);
            if (process is null)
            {
                return false;
            }

            process.StandardOutput.ReadToEnd();
            process.StandardError.ReadToEnd();
            process.WaitForExit();
            return process.ExitCode == 0;
        }
        catch (Exception e) when (e is System.ComponentModel.Win32Exception or IOException)
        {
            return false;
        }
    }
}

File: tests/units/PSWSMan.Authentication.Tests/acceptor.py
# Copyright: (c) 2026, Jordan Borean (@jborean93) <jborean93@gmail.com>
# MIT License (see LICENSE or https://opensource.org/licenses/MIT)

"""Authentication acceptor driven by the PSWSMan.Authentication.Tests project.

The tests spawn this script and talk to it over stdin/stdout, one JSON object
per line. It hosts a single pyspnego acceptor context so the module's client
side authentication and message protection can be verified against an
independent implementation without any HTTP or WSMan involved.

Every request is ``{"op": "...", ...}`` and every response is either
``{"ok": true, ...}`` or ``{"ok": false, "type": "<exception>", "error": "..."}``.
Binary values are base64 strings.

Operations:

``create``
    protocol, hostname, service, options (list of NegotiateOptions names),
    context_req (int, optional), channel_bindings (base64 application data,
    optional), tls (optional, CredSSP only: ``tls1.3``, ``tls1.2-aead`` or
    ``tls1.2-cbc`` pins the acceptor's TLS version and cipher family).
    Creates the acceptor context.
``step``
    token (base64 or null). Returns token (base64 or null) and complete.
``wrap`` / ``unwrap``
    data. Returns data and encrypted.
``wrap_winrm``
    data. Returns header, data and padding_length.
``unwrap_winrm``
    header, data. Returns data.
``query``
    Returns complete, negotiated_protocol, context_attr and client_principal.
    A CredSSP acceptor adds tls_protocol and tls_cipher once the handshake is
    done and delegated_credentials once it has received them.
``exit``
    Ends the process.
"""

from __future__ import annotations

import base64
import json
import os
import shutil
import ssl
import sys
import tempfile
import typing as t

import spnego
import spnego._ntlm
import spnego.tls
from spnego._ntlm_raw.messages import Challenge, NegotiateFlags, Version
from spnego.channel_bindings import GssChannelBindings


class _ChallengeWithVersion(Challenge):
    """Works around pyspnego <= 0.12.2 building a CHALLENGE without a Version field.

    The NTLM acceptor copies the initiator's flags into the CHALLENGE message,
    including NTLMSSP_NEGOTIATE_VERSION, but only writes the 8 byte Version
    structure when one is explicitly supplied. Strict decoders like gssntlmssp
    place the payload after that structure when the flag is set so every field
    offset then points 8 bytes too early and the message is rejected. Supplying
    a version whenever the flag is set produces the layout Windows sends.
    """

    def __init__(self, flags: int = 0, *args: t.Any, **kwargs: t.Any) -> None:
        if flags & NegotiateFlags.version and not kwargs.get("version"):
            kwargs["version"] = Version(major=10, minor=0, build=0, revision=0x0F)
        super().__init__(flags, *args, **kwargs)


spnego._ntlm.Challenge = _ChallengeWithVersion  # type: ignore[misc]


def b64(value: bytes | None) -> str | None:
    return base64.b64encode(value).decode() if value is not None else None


def unb64(value: str | None) -> bytes | None:
    return base64.b64decode(value) if value is not None else None


# The cipher lists are TLS 1.2 only, TLS 1.3 suites are all AEAD and cannot be restricted this way.
_TLS_VARIANTS: dict[str, tuple[ssl.TLSVersion, str | None]] = {
    "tls1.3": (ssl.TLSVersion.TLSv1_3, None),
    "tls1.2-aead": (ssl.TLSVersion.TLSv1_2, "ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384"),
    "tls1.2-cbc": (ssl.TLSVersion.TLSv1_2, "ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384"),
}


def create_tls_context(variant: str) -> spnego.tls.CredSSPTLSContext:
    """Builds a CredSSP acceptor TLS context pinned to one TLS version and cipher family.

    Mirrors what pyspnego does for its default acceptor context, a generated
    certificate loaded into default_tls_context, with the version and ciphers
    restricted on top.
    """
    version, ciphers = _TLS_VARIANTS[variant]

    ctx = spnego.tls.default_tls_context(usage="accept")
    cert_pem, key_pem, public_key = spnego.tls.generate_tls_certificate()

    # load_cert_chain opens the file itself which fails on Windows for a NamedTemporaryFile, use a directory.
    temp_dir = tempfile.mkdtemp()
    try:
        cert_path = os.path.join(temp_dir, "cert.pem")
        with open(cert_path, mode="wb") as fd:
            fd.write(cert_pem)
            fd.write(key_pem)
        ctx.context.load_cert_chain(cert_path)
    finally:
        shutil.rmtree(temp_dir)

    ctx.public_key = public_key
    ctx.context.minimum_version = version
    ctx.context.maximum_version = version
    if ciphers:
        ctx.context.set_ciphers(ciphers)

    return ctx


class Acceptor:
    def __init__(self) -> None:
        self.context: spnego.ContextProxy | None = None

    def _require_context(self) -> spnego.ContextProxy:
        if self.context is None:
            raise RuntimeError("The context has not been created")
        return self.context

    def op_create(self, request: dict[str, t.Any]) -> dict[str, t.Any]:
        options = spnego.NegotiateOptions.none
        for name in request.get("options") or []:
            options |= spnego.NegotiateOptions[name]

        bindings = None
        app_data = unb64(request.get("channel_bindings"))
        if app_data is not None:
            bindings = GssChannelBindings(application_data=app_data)

        kwargs: dict[str, t.Any] = {}
        if request.get("context_req") is not None:
            kwargs["context_req"] = spnego.ContextReq(request["context_req"])
        if request.get("tls") is not None:
            kwargs["credssp_tls_context"] = create_tls_context(request["tls"])

        self.context = spnego.server(
            hostname=request.get("hostname", "unspecified"),
            service=request.get("service", "host"),
            channel_bindings=bindings,
            protocol=request.get("protocol", "negotiate"),
            options=options,
            **kwargs,
        )
        return {}

    def op_step(self, request: dict[str, t.Any]) -> dict[str, t.Any]:
        context = self._require_context()
        out_token = context.step(unb64(request.get("token")))
        return {"token": b64(out_token), "complete": context.complete}

    def op_wrap(self, request: dict[str, t.Any]) -> dict[str, t.Any]:
        context = self._require_context()
        result = context.wrap(unb64(request["data"]) or b"", encrypt=request.get("encrypt", True))
        return {"data": b64(result.data), "encrypted": result.encrypted}

    def op_unwrap(self, request: dict[str, t.Any]) -> dict[str, t.Any]:
        context = self._require_context()
        result = context.unwrap(unb64(request["data"]) or b"")
        return {"data": b64(result.data), "encrypted": result.encrypted}

    def op_wrap_winrm(self, request: dict[str, t.Any]) -> dict[str, t.Any]:
        context = self._require_context()
        result = context.wrap_winrm(unb64(request["data"]) or b"")
        return {"header": b64(result.header), "data": b64(result.data), "padding_length": result.padding_length}

    def op_unwrap_winrm(self, request: dict[str, t.Any]) -> dict[str, t.Any]:
        context = self._require_context()
        data = context.unwrap_winrm(unb64(request["header"]) or b"", unb64(request["data"]) or b"")
        return {"data": b64(data)}

    def op_query(self, request: dict[str, t.Any]) -> dict[str, t.Any]:
        context = self._require_context()
        info: dict[str, t.Any] = {
            "complete": context.complete,
            "negotiated_protocol": context.negotiated_protocol,
            "context_attr": int(context.context_attr),
            "client_principal": context.client_principal,
        }

        # Both are CredSSP only, the TLS details are known once the handshake has finished.
        ssl_object = context.get_extra_info("ssl_object")
        cipher = ssl_object.cipher() if ssl_object is not None else None
        if cipher is not None:
            info["tls_cipher"], info["tls_protocol"], _ = cipher

        # Only a completed CredSSP acceptor returns this, the tests only delegate password credentials.
        password_creds = context.get_extra_info("client_credential")
        if password_creds is not None:
            info["delegated_credentials"] = {
                "domain": password_creds.domain_name,
                "username": password_creds.username,
                "password": password_creds.password,
            }

        return info

    def handle(self, request: dict[str, t.Any]) -> dict[str, t.Any]:
        op = request.get("op")
        handler = getattr(self, f"op_{op}", None) if isinstance(op, str) else None
        if handler is None:
            raise ValueError(f"Unknown operation '{op}'")

        return handler(request)


def main() -> None:
    acceptor = Acceptor()
    stdin = sys.stdin.buffer
    stdout = sys.stdout.buffer

    while True:
        line = stdin.readline()
        if not line:
            break

        request = json.loads(line)
        if request.get("op") == "exit":
            break

        try:
            response = acceptor.handle(request)
            response["ok"] = True
        except Exception as e:  # noqa: BLE001 - every failure is reported to the test
            response = {"ok": False, "type": type(e).__name__, "error": str(e)}

        stdout.write(json.dumps(response).encode() + b"\n")
        stdout.flush()


if __name__ == "__main__":
    main()
